Fixed paste bugs on headings and lists

jorgefilipecosta opened the pull request:
## Description
This PR aims to fix https://github.com/WordPress/gutenberg/issues/3241, https://github.com/WordPress/gutenberg/issues/3726, and https://github.com/WordPress/gutenberg/issues/3351. Affecting paste on headings and list blocks.
It proposes an alternative approach to https://github.com/WordPress/gutenberg/pull/3842, this alternative is generic and requires very low effort from block creators.


The idea behind this PR is that our definition of what HTML elements are inline elements should not be static. We should allow blocks to set HTML elements they would like to be treated as inline elements.
So list block can set list elements to be interpreted as inline, this way when we paste a list into an empty list block we will not create a new list as it happens right now. Also when we paste list items in the middle of list block we "merge" the items as it would be expected.

This PR misses the addition of test cases and updates in the documentation, they will be added as soon as we are more confident we can follow this approach.


## How Has This Been Tested?
Paste an HTML list in an empty list block see no new list was created and the list was correctly pasted.
Paste an HTML list in the middle of a list block see the list was correctly merged.
Paste an HTML heading block in an empty heading block see no new headings were created.
Paste an HTML heading block in a non-empty heading block see contents were correctly merged.

## After Screenshots:
### Empty list paste
![dec-12-2017 13-27-15](https://user-images.githubusercontent.com/11271197/33886973-32013fb6-df40-11e7-87ca-b22f69ec96ff.gif)
### Non-Empty list paste
![dec-12-2017 13-22-41](https://user-images.githubusercontent.com/11271197/33887320-5182584c-df41-11e7-967c-1cbafc58618b.gif)
### Heading paste
![dec-12-2017 13-29-55](https://user-images.githubusercontent.com/11271197/33887167-cc2ff384-df40-11e7-96f1-104ae63c891a.gif)


## Before Screenshots:
### Empty list paste
![dec-12-2017 13-26-21](https://user-images.githubusercontent.com/11271197/33887000-41a90098-df40-11e7-933c-2e0265b8f8cf.gif)
### Non-Empty list paste
![dec-12-2017 13-24-27](https://user-images.githubusercontent.com/11271197/33887313-49dec3be-df41-11e7-87b9-f91d7b90db23.gif)
### Heading paste
![dec-12-2017 13-30-40](https://user-images.githubusercontent.com/11271197/33887161-c66d6e86-df40-11e7-8fcd-d753dce3f256.gif)

jorgefilipecosta committed:
Implemented isInlineForTag mechanism

This mechanism allows treating elements as inline if they are being pasted in specific tags. So for example in heading blocks we can interpret headings as inline content and on lists we can interpret lists as inline content.

mcsf commented:
The author-facing benefit of this change is very clear, and welcome. However, could we go further and let `Editable` infer more? The required `tagName` prop seems like a pretty good hint:

- For **Heading**, `tagName` is expected to be one of `[ 'h1', 'h2', 'h3', 'h4', 'h5', 'h6' ]`, which exactly fits its `additionalInlineNodes` prop.
- For **List**, `tagName` is expected to be one of `[ 'ol', 'ul' ]`, which is almost the same as its `additionalInlineNodes` prop, save for the missing `'li'`. Still pretty close.

Maybe there are use cases out there that warrant a fully customizable prop like `additionalInlineNodes`, but is it worth exploring a more autonomous Editable?

iseulde commented:
I'm liking how this fixes the list problem as well!

I find `additionalInlineNodes` a bit poorly named as the tags that will be supplied are not inline tags/nodes at all. The name should communicate that the presence of these tags alone will not trigger automatically trigger block mode for paste, which is hard to describe in a name... Maybe something like `pasteAutoInlineModeTags`?

@mcsf Not sure how we could get rid of this prop by looking at the Editable `tagName`?

iseulde commented on blocks/api/raw-handling/index.js:
```diff
@@ -36,7 +36,7 @@ import shortcodeConverter from './shortcode-converter';
  * @return {Array|String}                     A list of blocks or a string, depending on `handlerMode`.
  */
-export default function rawHandler( { HTML, plainText = '', mode = 'AUTO' } ) {
+export default function rawHandler( { HTML, plainText = '', mode = 'AUTO', additionalInlineNodes = [] } ) {
```
This change needs to be documented above.

iseulde commented on blocks/editable/index.js:
```diff
@@ -72,6 +73,10 @@ function getFormatProperties( formatName, parents ) {
 
 const DEFAULT_FORMATS = [ 'bold', 'italic', 'strikethrough', 'link' ];
 
+const AUTO_INLINE_MODE_TAG_GROUPS = [
```
Since this is paste specific, I'm wondering if maybe we can't merge this with the whitelist into some sort of config for the raw handling. We'd just supply `rawHandler` with a base tag. Thoughts?

jorgefilipecosta replied:
So we would pass the tag to rawHandler and rawHandler would build the array using AUTO_INLINE_MODE_TAG_GROUPS right?
The only disadvantage I see in this approach is that we lose the possibility to pass a custom array in editable, maybe it will be useful for some block. But right now we don't have a use case for that so I'm fine with that change :)

mcsf replied:
I was also thinking this would live along with the whitelist stuff and `rawHandler` would only care about the one tag, yes.

jorgefilipecosta replied:
Hi @mcsf and @iseulde some changes were applied that move things in that direction. Feel free to a have a new look and let me your thoughts. Thank you for your reviews.

mcsf commented:
> @mcsf Not sure how we could get rid of this prop by looking at the Editable tagName?

@iseulde I was thinking of something along the lines of what the branch now looks like; does that answer your question?

iseulde commented:
@mcsf Yes :)

mcsf commented on blocks/api/raw-handling/utils.js:
```diff
@@ -63,8 +73,18 @@ export function isAttributeWhitelisted( tag, attribute ) {
+	return includes(
+		find( inlineWhitelistTagGroups, group => includes( group, tagName.toUpperCase() ) ),
+		nodeName,
+	);
```
Using `find` poses a problem if the same tag is found in multiple groups. Example:

```js
const groups = [
  [ 'a', 'b', 'c' ],
  [ 'c', 'd', 'e' ],
];

isInline( 'd', 'c' ); // false, but should be true
```

Furthermore, the time complexity is a lot greater than necessary because of the nested `includes > find > includes` algorithm, when actually **O(n)** is feasible.

Using `Array#some + lodash#includes` helps us reduce complexity and IMO improves readability, though `Array#some + for` would be the fastest, but also much more verbose. See [this working jsfiddle](https://jsfiddle.net/mcsf/88e08w2j/).

jorgefilipecosta replied:
Thank you for providing a fiddle 👍 And finding the logic bug, as before we used find to get an array and then passed the array I ported that logic but it is now not fitted. I used the some + includes as it is less verbose, and although less efficient we will not have many groups so that should not be noticeable.

mcsf commented on blocks/api/raw-handling/utils.js:
```diff
@@ -63,8 +73,18 @@ export function isAttributeWhitelisted( tag, attribute ) {
+}
+
+export function isInline( node, tagName ) {
+	return !! inlineWhitelist[ node.nodeName.toLowerCase() ] || nodeIsInlineOnSpecificTag( node.nodeName, tagName );
```
I suggest adding `const nodeName = node.nodeName.toLowerCase()` to simplify the inner function calls, as well as standardize on always using lowercase tag names in that module.

mcsf replied:
The name `nodeIsInlineOnSpecificTag` is longer but not necessarily clearer, because the entire problem requires explaining with more than a few words. :) Suggestions: `isInlineForTag`, `isInlineGivenTag`, `isInlineForBaseTag`, complemented with a jsdoc for that function.

mcsf commented on blocks/api/raw-handling/test/is-inline-content.js:
```diff
@@ -12,12 +12,14 @@ describe( 'stripWrappers', () => {
 		equal( isInlineContent( '<em>test</em><div>test</div>' ), false );
 		equal( isInlineContent( 'test<br><br>test' ), false );
 		equal( isInlineContent( '<em><div>test</div></em>' ), false );
+		equal( isInlineContent( '<li>test</li>', 'p' ), false );
```
I suggest adding a test that mixes the list group with the header group.

mcsf commented on blocks/api/raw-handling/index.js:
```diff
@@ -28,15 +28,16 @@ import shortcodeConverter from './shortcode-converter';
+ *                                             * 'AUTO': Decide based on the content passed.
+ *                                             * 'INLINE': Always handle as inline content, and return string.
+ *                                             * 'BLOCKS': Always handle as blocks, and return array of blocks.
+ * @param  {Array}         [options.tagName]   The tag where content will be inserted in.
```
→ "The tag into which content will be inserted."

jorgefilipecosta committed:
Added tag groups for headings and lists to the inlineWhitelistTagGroups mechanism.

This improves the behaviour of paste in already existing list and heading blocks.

jorgefilipecosta committed:
Added test cases to isInlineContent to take into account the new isInlineForTag mechanism.

jorgefilipecosta commented:
Hi @mcsf thank you for reviewing in detail this PR, it was updated.

mtias commented:
@jorgefilipecosta @iseulde @mcsf not sure when we regressed on this, but we can no longer paste a list like:

```
* One
* Two
* Three
```

and have it converted into a proper list.

iseulde commented:
@mtias This probably never worked with the list on its own. If there are no two consecutive line breaks detected, MarkDown parsing will be skipped to make sure inline paste works correctly. I'll attempt to fix this right now. Maybe we can just check for a single line break or try to get rid of this check entirely.
